flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CompletedCheckpointStore.java

```diff
@@ -55,31 +54,26 @@
 	 * added.
 	 */
 	default CompletedCheckpoint getLatestCheckpoint(boolean isPreferCheckpointForRecovery) throws Exception {
-		if (getAllCheckpoints().isEmpty()) {
+		List<CompletedCheckpoint> allCheckpoints = getAllCheckpoints();
+		if (allCheckpoints.isEmpty()) {
 			return null;
 		}
 
-		CompletedCheckpoint candidate = getAllCheckpoints().get(getAllCheckpoints().size() - 1);
-		if (isPreferCheckpointForRecovery && getAllCheckpoints().size() > 1) {
-			List<CompletedCheckpoint> allCheckpoints;
-			try {
-				allCheckpoints = getAllCheckpoints();
-				ListIterator<CompletedCheckpoint> listIterator = allCheckpoints.listIterator(allCheckpoints.size() - 1);
-				while (listIterator.hasPrevious()) {
-					CompletedCheckpoint prev = listIterator.previous();
-					if (!prev.getProperties().isSavepoint()) {
-						candidate = prev;
-						LOG.info("Found a completed checkpoint before the latest savepoint, will use it to recover!");
-						break;
-					}
+		CompletedCheckpoint lastCompleted = allCheckpoints.get(allCheckpoints.size() - 1);
+
+		if (lastCompleted.getProperties().isSavepoint() && isPreferCheckpointForRecovery && allCheckpoints.size() > 1) {
+			ListIterator<CompletedCheckpoint> listIterator = allCheckpoints.listIterator(allCheckpoints.size() - 1);
```

Comment:
You can simplify code further: as I've mentioned before, you don't need a special `lastCompleted.getProperties().isSavepoint()` check, instead instantiate the iterator from the very last element `allCheckpoints.listIterator(allCheckpoints.size())`.

The whole method can be structured as:
```
if (allCheckpoints.isEmpty()) {
    return null;
}
if (isPreferCheckpointForRecovery) {
    // try to find the latest checkpoint by reverse traversal over allCheckpoints
    // return it if found
}
// return the latest (be it checkpoint or savepoint)
return allCheckpoints.get(allCheckpoints.size() - 1);
```

Reply:
This way it's easy to give a log message when we fall back to a checkpoint

Reply:
@gyfora Yes. My last concern is also about logging. In your version given the last completed is a checkpoint we actually log nothing :P

Reply:
I would only log if we don't follow the standard fallback-to-last approach

Reply:
thanks for your explanation :-)

Reply:
If you want to log situations when another (non last element) is picked, you could have added additional check by reference that the picked element is not the last one.

If you are going to keep your version, could you please move `lastCompleted.getProperties().isSavepoint()` after `isPreferCheckpointForRecovery`, so it's not evaluated eagerly.